XCode 12 support

cheif opened the pull request:
### What does this PR do
Makes sure that you can build with Xcode 12

This PR makes two changes, first we upgrade `realm-cocoa` to v5.5.0, which has works with XCode 12.
The other thing is adding `carthage.sh` from https://github.com/Carthage/Carthage/blob/master/Documentation/Xcode12Workaround.md to work around the current build issues with carthage.

So, with this PR you should be able to build by running `carthage.sh` instead of `carthage`, AKA running something like: `./carthage.sh bootstrap --platform tvOS --project-directory ./ --cache-builds` for tvOS builds.

This should solve #1437, although I think we should keep it open until this is fixed in upstream carthage, seems like they are working on it https://github.com/Carthage/Carthage/issues/3019

Fixes #1432
Fixes #1437
Fixes #1445 

### How should this be manually tested
I've launched it, but hasn't had time to try it out a lot, the realm update might break something, but I think that it should be fairly safe. But more testing would of course be nice

Theiof commented:
Hai!

Unfortunately, after 10 years and 10 million spam, I do not use this email address any more.

If you wish to contact me or ask for my new email address, reach out to me on Twitter: @dedanieldeboer

Wish you all the best,

De Daniel de Boer

JoeMatt commented:
We should leave the Rome part of azure. I'll eventually fix it. It speeds up builds a ton.

cheif commented:
> We should leave the Rome part of azure. I'll eventually fix it. It speeds up builds a ton.

Couldn't we just re-add it when rome is fixed, to avoid clutter in the pipeline?
Anyway, I just removed it to try to make sense of what's happening on azure, and why it's failing. So I have no problem re-adding it either, since that's not what's causing problems.

cheif commented:
Ok, I can't seem to get this to work on azure. Right now it looks like the `carthage bootstrap` step uses the Swift 5.3 compiler, but the `xcodebuild` uses Swift 5.3.1, thus getting mismatches on dependencies vs. the actual app. This is nothing that we control, and I guess this is something that we'll have to wait for Azure to fix.

I've verified that this branch builds properly with Xcode 12.2 for both iOS and tvOS locally though, so I think that it's safe to merge this.

@JoeMatt Tell me if you'd like me to drop the last commit (removing the rome-stuff), to me it doesn't matter if we keep it or remove it, only that removing it makes the pipeline a bit cleaner right now.

urherenow commented:
I have cloned and built cheif/xcode-12 and it does build. freshly importing roms into the import folder work fine, and you can play them. but going into the menu and "Refresh Library" causes an instant crash. From then on, none of your previously imported roms work.

jasarien commented:
> I have cloned and built cheif/xcode-12 and it does build. freshly importing roms into the import folder work fine, and you can play them. but going into the menu and "Refresh Library" causes an instant crash. From then on, none of your previously imported roms work (the ROM file for this game cannot be found), even though there are all still there. I did this on an iPad Pro.

I think we should remove this “feature”. It was added very early when the library and import process were very different to how they are now.

urherenow commented:
So the real issue seems to be that the database (I'm guessing it's part of realm) isn't properly updated/managed with roms already located in the correct spot. To fix without deleting the default.realm stuff AND the roms, I have to first delete the troubled rom(s) from the main screen, then use whatever method to drop them into the Imports folder. I also must put .bin/.que files there (or a zip of them both) and allow it to go to the Conflicts folder, so that I can resolve the conflict in the main screen by picking the system. dropping it into the correct place to begin with, is broken.

Since I've been deployed for ~7 months, and have already updated my mac vm/xcode, I am unable to test if this is a new issue related to the xcode 12 fixes or if this issue crept into the main development branch while I was away. I hope I've provided enough information for you to check it out before the merge.

cheif commented:
I've managed to reproduce the crash at least, and it seems to be due to us reading stuff from realm as they are being deleted, I'm not sure if this is the result of a change in realm or if it's on master as well. Haven't got more time to look into it now though, but I think we should be able to fix it on this branch at least.

cheif commented:
Now the crash is fixed, although the refresh-library functionality seems to be pretty broken. The only thing happening is that we clear the database, and then on the next app-start we search the system-folders for roms etc. at least it's not crashing now.

cheif commented:
And now I managed to get Azure working as well, including using the Azure-cache for carthage stuff

JoeMatt commented:
The library issues, if I'm following this topic correctly (I scanned), comes down to 2 things,

1. Bugs in Realm thread race conditions vs the main thread (maybe a realm update fixes this but this bug has existed for a while).
2. Sometimes imports don't trigger a redraw. (T/his shouldb't crash though the UI will be out of sync until you close the library screen and reopen it)

The library VC needs a rewrite. it's one of the oldest parts of the code. It needs to be 

a) refactored, split
b) reactive UI updates using combine or RX. There's too many moving parts for delgate updates it's a bitch.

JoeMatt commented on PVLibrary/PVLibrary/RealmPlatform/PVObject.swift:
```diff
@@ -21,10 +21,6 @@ public extension PVObject where Self: Object {
         try RomDatabase.sharedInstance.add(self, update: update)
     }
 
-    static func deleteAll() throws {
```
Why remove this?

JoeMatt commented on PVLibrary/PVLibrary/RealmPlatform/RomDatabase.swift:
```diff
@@ -374,19 +374,12 @@ public extension RomDatabase {
         }
     }
 
-    func deleteAll<T: Object>(_ type: T.Type) throws {
```
Again, why delete? Did realm remove deleteAll?

JoeMatt commented on Provenance/PVAppDelegate.swift:
```diff
@@ -97,6 +97,14 @@ final class PVAppDelegate: UIResponder, UIApplicationDelegate {
             libraryUpdatesController.addImportedGames(to: CSSearchableIndex.default(), database: RomDatabase.sharedInstance).disposed(by: disposeBag)
         #endif
 
+        // Handle refreshing library
```
Not sure how I feel about this living in AppDelegate.

JoeMatt reviewed (commented):
This looks good other than my inline comments.

If these are sensible, I can approve, or if you want to make a separate PR with an explanation of changes.

Thanks!

cheif commented:
@JoeMatt Yeah, you're correct, I moved the crash-fix here: https://github.com/Provenance-Emu/Provenance/pull/1448 instead, so this PR should be the minimum to get XCode12 working now.

cheif commented:
> The library issues, if I'm following this topic correctly (I scanned), comes down to 2 things,
> 
> 1. Bugs in Realm thread race conditions vs the main thread (maybe a realm update fixes this but this bug has existed for a while).
> 2. Sometimes imports don't trigger a redraw. (T/his shouldb't crash though the UI will be out of sync until you close the library screen and reopen it)
> 
> The library VC needs a rewrite. it's one of the oldest parts of the code. It needs to be
> 
> a) refactored, split
> b) reactive UI updates using combine or RX. There's too many moving parts for delgate updates it's a bitch.

I actually think that 2 is working fine now, I did one rewrite where I moved most of the logic into `PVGameLibrary`, which then just exposes a stream of systems that's used in the ViewController, but of course it can still be improved a lot.

JoeMatt commented:
Awesome work thanks @cheif .

I'm having Carthage issues locally (hanging on clone or build), I only tried it late last night.
I'm assuming it's my machine since Azure is working.

cheif commented:
@JoeMatt did you use the new carthage.sh script, that's required for carthage to work with XCode12

jasarien commented:
Builds fine for me on Big Sur with Xcode 12.2 using the `carthage.sh` script. Gonna merge.
